File: PYTHON/applications/sci_rehab/measurement.py
"""
Observation model for SCI rehabilitation state estimation.

Ported from Katie's CODES/KATIE'S CODE/.../Measurement.m.

At each rehab session, only the two DYNAMIC states are observed:
    y = [SCIM_norm; BBS_norm] + G @ v * sqrt(dt)
(AIS, Age, DPI, Caregiver support are fixed patient attributes, known
at admission, never observed through a noisy channel here.) G is a
constant diagonal instrument-noise scaling -- SCIM has higher
inter-rater reliability (lower noise) than BBS.

Matches forward_pass.py's `measurement` callable contract exactly:
    measurement(dt, xa, u, constants, v, noise_index, augment_states,
                v_from_filter) -> (ny, K) array
noise_index=False -> the real (noisy) observation g(x) + G v sqrt(dt).
noise_index=i (1-indexed, matching MATLAB) -> the i-th column of the
measurement-noise diffusion matrix G, broadcast across all K columns --
this is what forward_pass.py differentiates to build Dx/Du, and comes
out identically zero there since G doesn't depend on x or u.

DEVIATION FROM THE LITERAL MATLAB SOURCE: `u`, `constants`, and
Measurement.m's broadcasting of `u` to match xk's column count are all
dead in the MATLAB source -- g depends only on xk, never on u or the
constants struct. Kept as unused parameters here for signature fidelity
with the shared measurement() contract (same convention as backward_pass.py's
unused `E`).
"""

import numpy as np

_G = np.diag([0.03, 0.05])  # [SCIM noise scale, BBS noise scale], (ny, nv)


def measurement(dt, xk, u, constants, v, noise_index, augment_states,
                  v_from_filter):
    Np1 = xk.shape[1]

    if noise_index:  # 1-indexed request for the noise-diffusion column
        col = _G[:, noise_index - 1]
        return np.tile(col[:, None], (1, Np1))

    g = xk[:2, :]  # observe SCIM (x1), BBS (x2) directly

    if not v_from_filter:
        noise = (_G @ v) * np.sqrt(dt)
    else:
        noise = v

    return g + noise

File: CODES/KATIE'S CODE/Curiosity_Algorithm_w_SINDy_V7_SGBA_personas/Measurement.m
function yk = Measurement(dt, xk, u, c, v, G_only_col, augment_states, v_from_filter)
%MEASUREMENT  Observation function for SCI rehabilitation state estimation.
%
%  In rehabilitation, we observe functional outcome scores at each session.
%  Observations: [SCIM_norm; BBS_norm] — the two dynamic states.
%  (AIS and other static attributes are known at admission and fixed.)
%
%  Measurement model: y = g(x) + G*v*sqrt(dt)
%  where G encodes measurement noise scaling per instrument.

%% ── Observation function g(x) ───────────────────────────────────────────

% Observe SCIM (x1) and BBS (x2) directly
g = [xk(1, :);   % SCIM_norm
     xk(2, :)];  % BBS_norm

Np1 = size(xk, 2);
if size(u, 2) == 1 && Np1 > 1
    u = u * ones(1, Np1);
end

ny = size(g, 1);   % 2 observations
nv = size(v, 1);

%% ── Noise ────────────────────────────────────────────────────────────────

if v_from_filter == 0
    for j = 1:Np1
        % Measurement noise scaling:
        %   SCIM has higher inter-rater reliability (lower noise) than BBS
        G(:, :, j) = diag([0.03, 0.05]);   % [ny x nv]
        Gv(:, j)   = G(:, :, j) * v(:, j);
    end
    noise = Gv * sqrt(dt);
else
    noise = v;
end

yk = g + noise;

if G_only_col > 0
    yk = G(:, G_only_col, :);
end

end
